doc/design/api.md

```diff
@@ -0,0 +1,277 @@
+# PaddlePaddle Design Doc
+
+## Ingredients
+
+As the first step of our design, we list important concepts in deep
+learning and try to figure their relationship, as shown below:
+
+```
+Model = {topology, parameters}
+
+Evaluator = {Model*, activations}
+- forward
+- test(cost, ...)
```

Comment:
@reyoung Good comments! For the "add `begin` and `end`" part, I don't think `begin` is necessary? Because the dependency is fully specified when we want to calculate `end`.